core/vm/contracts.go

```diff
@@ -540,6 +580,23 @@ func (c *blake2F) Run(input []byte) ([]byte, error) {
 	return output, nil
 }
 
+// epoch returns the current epoch, implemented as a native contract
+type epoch struct{}
+
+// RequiredGas returns the gas required to execute the pre-compiled contract.
+//
+// This method does not require any overflow checking as the input size gas costs
+// required for anything significant is so high it's impossible to pay for.
+func (c *epoch) RequiredGas(input []byte) uint64 {
+	return GasQuickStep
+}
+
+func (c *epoch) Run(input []byte) ([]byte, error) {
+	// Note the input was overwritten with the epoch of the current block
+	// So just format and return
+	return common.LeftPadBytes(input, 32), nil
```

Comment:
Is this format easily readable in the smart contract? Any example code for calling epoch precompile?

Reply:
This is exactly the same setup as the VRF in #3703, but the left padding is added to guarantee human readability. Without the left padding, the epoch 17 shows up as `0x1700000......` while with it, we get `0x000.....17`.
```solidity
function epoch() public view returns (bytes32 epochNumber) {
        bytes32 input;
        assembly {
            let memPtr := mload(0x40)
            if iszero(staticcall(not(0), 0xfb, input, 32, memPtr, 32)) {
                invalid()
            }
            epochNumber := mload(memPtr)
        }
    }

function _epoch() public view returns (uint256 epochNumber) {
        epochNumber = uint256(epoch());
    }
```

Reply:
The below `go` code should hopefully demonstrate why the padding is needed, and helpful. 
 - As the epoch gets larger, it doesn't fit in a single byte and the contract therefore needs to specify how many bytes it is expecting (32 in this case). 
 - Converting those bytes into a `uint256` in Solidity is easier if those are left padded with zero, instead of right (the default, if there is no padding provided since it reads from `memory` as used in [core/vm/instructions.go#L773](https://github.com/harmony-one/harmony/blob/f95daff0b775aa4202e42993c175d2be10eca721/core/vm/instructions.go#L773).)
```go
package main

import (
	"fmt"
	"math/big"

	"github.com/ethereum/go-ethereum/common"
)

func main() {
  x := big.NewInt(841)
  y := common.LeftPadBytes(x.Bytes(), 64)
  fmt.Println(x.Bytes())
  fmt.Println(y)
}
```

```
[3 73]
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 3 73]
```